File: test-suite/elixir/clean/open_redirect.ex
defmodule CleanOpenRedirect do
  import Phoenix.Controller
  import Plug.Conn

  @allowed_redirect_hosts MapSet.new(["app.example.com", "accounts.example.com"])

  def safe_redirect_url(raw) do
    if String.starts_with?(raw, "/") and not String.starts_with?(raw, "//") do
      raw
    else
      uri = URI.parse(raw)

      unless uri.scheme == "https" and MapSet.member?(@allowed_redirect_hosts, uri.host) do
        raise ArgumentError, "blocked redirect"
      end

      raw
    end
  end

  def helper_validated_query(conn) do
    target = safe_redirect_url(conn.params["next"] || "/")
    redirect(conn, to: target)
  end

  def helper_validated_header(conn) do
    target =
      conn
      |> Plug.Conn.get_req_header("x-redirect-url")
      |> List.first()
      |> safe_redirect_url()

    Phoenix.Controller.redirect(conn, external: target)
  end

  def helper_validated_query_params(conn) do
    target = safe_redirect_url(Map.get(conn.query_params, "return_to") || "/")
    redirect(conn, external: target)
  end

  def inline_local_guard(conn) do
    target = conn.params["continue"] || "/"

    unless String.starts_with?(target, "/") and not String.starts_with?(target, "//") do
      raise ArgumentError, "blocked redirect"
    end

    redirect(conn, to: target)
  end

  def inline_host_allowlist(conn) do
    target = conn.params["return_url"] || "https://app.example.com/dashboard"
    uri = URI.parse(target)

    unless uri.scheme == "https" and MapSet.member?(@allowed_redirect_hosts, uri.host) do
      raise ArgumentError, "blocked redirect"
    end

    redirect(conn, external: target)
  end

  def validated_location_header(conn) do
    location = safe_redirect_url(conn.params["location"] || "/")

    conn
    |> put_resp_header("location", location)
    |> send_resp(302, "")
  end
end

File: test-suite/rust/clean/open_redirect.rs
use axum::extract::Query;
use axum::response::Redirect;
use std::collections::HashMap;
use url::Url;

const LOCATION: &str = "Location";

struct HeaderMap;
struct HeaderValue;
struct Request;
struct Response;
struct ResponseBuilder;

impl HeaderMap {
    fn get(&self, _name: &str) -> Option<HeaderValue> {
        Some(HeaderValue)
    }
}

impl HeaderValue {
    fn to_str(&self) -> Result<&str, ()> {
        Ok("/dashboard")
    }
}

impl Request {
    fn headers(&self) -> HeaderMap {
        HeaderMap
    }
}

impl Response {
    fn builder() -> ResponseBuilder {
        ResponseBuilder
    }
}

impl ResponseBuilder {
    fn header(self, _name: &str, _value: impl AsRef<str>) -> Self {
        self
    }
}

fn allowlist_redirect_url(raw: &str) -> Result<String, &'static str> {
    if raw.starts_with('/') && !raw.starts_with("//") {
        return Ok(raw.to_string());
    }

    let parsed = Url::parse(raw).map_err(|_| "invalid redirect")?;
    if parsed.scheme() != "https" {
        return Err("blocked redirect scheme");
    }

    let allowed_redirect_hosts = ["app.example.com", "accounts.example.com"];
    if !allowed_redirect_hosts.contains(&parsed.host_str().unwrap_or_default()) {
        return Err("blocked redirect host");
    }

    Ok(raw.to_string())
}

fn redirect_with_safe_helper(Query(params): Query<HashMap<String, String>>) -> Redirect {
    let raw = params.get("next").map(String::as_str).unwrap_or("/");
    let target = allowlist_redirect_url(raw).unwrap_or_else(|_| "/".to_string());
    Redirect::to(&target)
}

fn redirect_with_inline_local_guard(Query(params): Query<HashMap<String, String>>) -> Result<Redirect, &'static str> {
    let target = params.get("continue").cloned().unwrap_or_default();
    if !(target.starts_with('/') && !target.starts_with("//")) {
        return Err("blocked redirect");
    }
    Ok(Redirect::temporary(&target))
}

fn location_header_with_safe_helper(req: Request) -> ResponseBuilder {
    let raw = req
        .headers()
        .get("x-return-to")
        .and_then(|value| value.to_str().ok())
        .unwrap_or("/");
    let location = allowlist_redirect_url(raw).unwrap_or_else(|_| "/".to_string());
    Response::builder().header(LOCATION, location)
}
